python/phase-2.md

```diff
@@ -0,0 +1,830 @@
+  * Instead, like Java and Go, we move that to the package level
+* Should we move `is_replaying` to a place that makes it more clear it's unsafe?
+* It's clearer to require uses of `temporalio.workflow.now()` than to wait for and require overriding `datetime.now()`
+  when a deterministic sandbox appears. It is clearer to disallow deterministic calls than implicitly replace some.
+* `logger` uses a special adapter that is skipped during replay
+
+#### `temporalio.workflow.CancellationScope`
+
+Rough sketch:
+
+```python
+class CancellationScope:
+    @property
+    @staticmethod
+    def current() -> CancellationScope:
+        pass
+
+    def __init__(
+        self,
+        *,
+        # Default is current
+        parent: Optional[CancellationScope] = None,
+        detached: bool = False,
+        timeout: Optional[timedelta] = None,
+    ) -> None:
+        pass
+
+    def _push(self) -> None:
+        pass
+
+    def _pop(self) -> None:
+        pass
+
+    await def run(self, fn: Callable[..., Union[Awaitable[T], T]]) -> T:
+        pass
+
+    def __enter__(self) -> CancellationScope:
+        pass
+
+    def __exit__(self) -> None:
+        pass
+
+    def cancel(self) -> None:
+        pass
+
+    @property
+    def cancelled(self) -> bool:
+        pass
+```
+
+Notes:
+
+* Unlike TypeScript, Python still allows you to explicitly cancel individual activities and child workflows if you
+  started them with `start_`
+  * The consistency of having everything that has a `start_` be cancellable has value. I still expect most will use the
+    `execute_` form though
+* The workflow will have a root scope
+* This will work using `with CancellationScope()` and similar and is powered by `contextvars`
+  * `run` is async because the callable may have an async result, but `with` is the preferred approach for explicit use
+* `start_activity`/`execute_activity`/`start_child_workflow`/`execute_child_workflow` will implicitly use the current
+  cancellation scope
+* Do we need extra capabilities such as:
+  * Ability to explicitly add cancellables?
+  * Ability to unset the current instead of just forcing a detached one each time you want to opt out?
+* Should we disallow the `parent` to even be set in the constructor?
+  * Thinking we probably shouldn't allow you to set that
+* Cancellation of a workflow or a scope causes all things started with that to be issued a cancel
+  * Originally we weren't gonna use cancellation scopes but instead reuse the `asyncio` cancel feature but that throws
+    inside of awaited things and we want to cancel whether awaited or not
+  * This means `asyncio` cancellation (and `shield` etc) affects only in-workflow tasks awaited on, _not_ the actual
+    underlying activity/child being executed
+    * We could disable these features of `asyncio` so people aren't confused by their concept of cancellation and ours,
+      but there is still value in cancelling just the tasks
```

Comment:
What's the value in cancelling the tasks? I'd be inclined to say just remove this potential bit of confusion, but, since it will break at runtime maybe that's of dubious value anyway. Seems like a footgun people will run into but I dunno that there's any easy fix